### PR TITLE
Disallow aliases of functor arguments in module types

### PR DESCRIPTION
**Pull request by gasche**
This PR is a proposed fix for #11441 and #11442, co-written with @clementblaudeau.

For signatures of the form

```ocaml
module F(M : S) =
  ...
  module type T = sig
    module N = M
  end
end
```

we weaken `module N = M` into `module N : S`, following the logic in the typing of structures.

Another possible approach would be to check at functor-application time to remove aliases to unnamed arguments; this may accept more definitions but it is different from the approach taken for aliases of functor arguments in module expressions.

cc @garrigue, @lpw25: what do you think?

### CODE REVIEW

**Comment by lpw25** [comment]
I think this is not a good idea. It will be confusing to users to get a type unequal to the one they asked for. The reason it works for structures is that those are in a covatiant position so it is fine to give a weaker type. Module types are invariant so cannot be arbitrarily weakened.

This change also won't prevent people creating signatures with aliases to functor parameters via other means, which I suspect is impossible to prevent.

**Comment by clementblaudeau** [comment]
Inside the body of a functor, the user cannot build a module that the typechecker accepts as an alias to the argument, so a module type featuring such alias is inhabited. The only use cases are thus when the alias is substituted by an actual named argument at functor application, right ? Can't this be used to have aliases leaking through a functor which poses a problem for compilation ? 

> It will be confusing to users to get a type unequal to the one they asked for. 

That's true, we could also have the typechecker refuse such an alias (inside of a module type) instead of silently removing it. It is linked with #11443

**Comment by gasche** [comment]
@lpw25 which approach would you recommend to fix these issues? Forbid these programs instead of weakening the signature, as Clément suggests?

**Comment by garrigue** [comment]
I suppose it should just be illegal, i.e. the user should get an error.
The syntax `module N = M` in a signature indicates explicitly an alias, so it would be strange to give it another meaning.

Doing it at functor application time is interesting too, but it would still require path ascription, which are not available yet I believe.

**Comment by lpw25** [comment]
Having it be an error would be fine. Although it won't actually stop you from making modules with such signatures, which I don't think can actually be achieved. For example,

```ocaml
module type Type = sig module type S end;;
module type Type = sig module type S end

# module Foo (A : Type) (X : A.S) = struct module X = X end;;
module Foo : functor (A : Type) (X : A.S) -> sig module X : A.S end

# module Bad = Foo(struct module type S = sig module M : sig end module N = M end end);;
module Bad :
  functor (X : sig module M : sig end module N = M end) ->
    sig module X : sig module M : sig end module N = M end end
```

As I've advocated for a long time, the only way to fix this mess is to add proper support for aliases to functor parameters and by extension functor applications. See discussions in #10435, #10612 and #10673. I believe this could be achieved in the following way:

1. Add `module M == N` as syntax for absent module aliases, only supported when `N` is a compilation unit or itself an absent module alias. Make other aliases present by default. Have `-no-alias-deps` cause aliases to compilation units in covariant positions to automatically be made absent. Support coercions from absent aliases to present aliases but not the other way around.
2. Add support for present aliases to functor applications.
3. Add support for transparent module ascription. Implemented as an optional signature on present aliases.
4. Add support for aliases to functor arguments. When applying functors substitute the transparently ascribed module for the parameter -- which will only affect module aliases involving the parameter.
5. (optional) Add an option `-pure-app-funct` that requires all applicative functors be pure. This can use a variation of the value restriction for this check, or another effect mechanism is someone has one lying around.

The only other direction that seems plausible to me is some kind of modal system to track which module paths are actually pure/aliasable, but I'm pretty sure such system would be just as complicated as the one I propose as well as more confusing and probably less expressive.

**Comment by gasche** [comment]
Sorry for being slow on this, but I don't understand your example, what is it demonstrating? I don't see an alias involving a functor argument in the signature of `Bad`.

I believe that there is consensus that adding transparent ascription and using them to enable module aliases on functor arguments would be nice. But it seems that no one has volunteered the effort to implement this yet, and it is a non-trivial amount of work. (In comparison the change in the present PR is tiny, it merely tries to fix a small issue with the current specification.)

**Comment by lpw25** [comment]
Sorry, I tried to shorten the example and it stopped being broken. Here is a version that actually has an alias to a functor parameter:
```ocaml
# module type Type = sig module type S end;;
module type Type = sig module type S end

# module type Arg = sig module M : sig end module type S = sig module N = M end end;;
module type Arg =
  sig module M : sig end module type S = sig module N = M end end
  
# module Foo (A : Type) (A_to_Type : A.S -> Type) (X : A.S) (Y : A_to_Type(X).S) = struct
    module Y = Y
  end;;
    module Foo :
  functor (A : Type) (A_to_Type : A.S -> Type) (X : A.S) (Y : A_to_Type(X).S)
    -> sig module Y : A_to_Type(X).S end
    
# module Bad = Foo(struct module type S = Arg end)(functor (X : Arg) -> struct module type S = X.S end);;
module Bad : functor (X : Arg) (Y : X.S) -> sig module Y : X.S end
```

> But it seems that no one has volunteered the effort to implement this yet

We have quite a lot of the required work from @mrmr1993 in those PRs, but he decided that he didn't like the interaction with impure applicative functors and we reached an impasse. If anyone would like to pick up that work again I would be happy to review and advise.

> the present PR is tiny, it merely tries to fix a small issue with the current specification.

I don't object to merging the version that emits an error. I just want to be clear that this will not provide any additional guarantees about what sorts of module types can be created.

**Comment by gasche** [comment]
Thanks for the example. It is bad because `module Y : X.S` in the output signature of the functor in fact contains an alias on a sub-module of `X`.

Here is a reproduction case for #11441 using a simplification of your example:

```ocaml
module Bad =
  functor
    (X : sig
      module M : sig end
      module type Alias = sig module N = M end
    end)
    (Y : X.Alias)
  -> struct
    module Y = Y
  end;;
(*
module Bad :
  functor
    (X : sig module M : sig end module type Alias = sig module N = M end end)
    (Y : X.Alias) -> sig module Y : sig module N = X.M end end
*)
(* Notice that the inferred signature for Y contains an alias on a submodule of X. *)

module Test = Bad(struct module M = struct end module type Alias = sig module N = M end end);;
(*
>> Fatal error: nondep_supertype not included in original module type
Fatal error: exception Misc.Fatal_error
*)
```

The present PR does not make this crash go away.

My understanding is that:
- our patch deals with signatures that contain aliases on a functor argument
-

**Comment by gasche** [comment]
@clementblaudeau  and myself just found a worrying variant of this category of examples, to "easily" create an alias on a functor argument:

```ocaml
module type T = sig end
module F = functor (Y : sig module X : T end) (Alias : sig module X' = Y.X end) -> Alias;;
(*
module F :
  functor (Y : sig module X : T end) (Alias : sig module X' = Y.X end) ->
    sig module X' = Y.X end
*)
``

**Comment by clementblaudeau** [comment]
Which can be made even simpler:
```ocaml
module type T
module F =
  functor (Y : T) (Alias : sig module X' = Y end) -> Alias;;
(* module F :
  functor (Y : T) (Alias : sig module X' = Y end) -> sig module X' = Y end *)
```

**Comment by gasche** [comment]
I just pushed a new commit that neutralizes aliases inside module types in functor arguments, similarly to what is done for aliases inside modules: `M = N` was forgotten/weakened into `M : S` in `functor (Arg : sig module M : S module M = N end) -> ...`, now this weakening also happens with `functor (Arg : sig module M : S module type T : (sig module M = N end) end) -> ...`.

I understand that @lpw25's recommendation is rather to fail instead of weaken in module types, and this of course makes sense. @clementblaudeau and myself are planning to tweak the PR to error as suggested. (But I will be in holidays in August, so there may be a slow down.)

The PR as-is already neutralizes all the "bad" examples shown in this thread so far. In particular, Leo's example is neutralized because the application `Foo(struct module type S = Arg end end)` returns a higher-order functor expecting an argument of type `Arg -> Type`, but the users cannot write a functor with a compatible type anymore (the aliases in `Arg` get weakened).

This being said, we understand that this PR (with either weakening or erroring on aliases on signatures) is does not fully solve the issue, because abstract signatures let us avoid this weakening/erroring logic by revealing the aliases only at functor-application time. For example, even with this PR we have:

```ocaml
module type Type = sig module type S end;;

module Foo (Sig : Type) (X : Sig.S) = X;;

module F = Foo(struct module type S = sig module M : sig end module N = M end end);;
(*
module F :
  functor (X : sig module M : sig end module N = M end) ->
    sig module M : sig end module N = M end
*)
```

which shows that we can still build those "forbidden" signatures.

To summarize:
- we propose to adapt the (imperfect) logic currently applied on structures also on signatures,
  but to fail with an error on non-aliasable aliases instead of weakening them
- one should also check functor applications to make the problem go away, but we are not planning to work on this (this is outside the mandate of "easy, simple fix to the current design", and probably investing efforts into allowing aliases on functor arguments is better)

**Comment by clementblaudeau** [comment]
This PR has been rendered obsolete by #13192, which throws an error when trying to build signatures with invalid aliases.

**Comment by gasche** [comment]
Closing, then. Thanks for the triaging help!
